### ISSUE TITLE
Lotus + Rack::Chunked streaming

### ISSUE THREAD

**Issue by berfarah** [lotus/lotus#303]
`Rack::Chunked` streaming does not work out of the box. There are two issues here.

## 1. Lotus' rendering policy
`Lotus::RenderingPolicy#render` wraps any non-nil body in an array. In order for `Rack::Chunked` to work, existing enumerables should not be wrapped. I monkey-patched this by changing [line 30](https://github.com/lotus/lotus/blob/master/lib/lotus/rendering_policy.rb#L30) of the rendering policy to leave the response if `body.respond_to?(:each)`.

On [line 69](https://github.com/lotus/lotus/blob/master/lib/lotus/rendering_policy.rb#L69) the rendering policy checks if the body is `#empty?` - this won't work for `Rack::Chunked::Body`, since it doesn't have an `empty?` method by default, so I had to add one in.

## 2. Lotus Server
I actually haven't investigated this at all, so I can't give a good cause. I was thrilled to get my code working with `rackup` so I'm probably going to move forward with that. That said, the above code is not sufficient to get `Rack::Chunked` working with `lotus server`

---
I have a gist outlining how I got `Rack::Chunked` to work: https://gist.github.com/berfarah/6cae66fb4ffe9e383cb5

**Comment by jodosha** [comment]
@berfarah Hello, have you tried to add `::Rack::Chunked` to the entire application middleware stack or to `config.ru`?

**Comment by berfarah** [comment]
@jodosha I'll revisit this now.

**Comment by berfarah** [comment]
I've committed sample applications [here](https://github.com/berfarah/lotus-streaming)

There are four branches:
* [config.ru middleware](https://github.com/berfarah/lotus-streaming/tree/configru) (doesn't work)
* [config.ru with monkey patching](https://github.com/berfarah/lotus-streaming/tree/monkeypatch-configru) (works with `rackup -s puma` but not `lotus s`
* [application.rb middlware](https://github.com/berfarah/lotus-streaming) (doesn't work)
* [application.rb with monkey patching](https://github.com/berfarah/lotus-streaming/tree/monkeypatch-application) (works with `rackup -s puma` but not `lotus s`

**Comment by pascalbetz** [comment]
Looked into it. Created a sample application as suggested by @berfarah. Application is at https://github.com/pascalbetz/lotus-streaming and the "hacks" are at https://github.com/pascalbetz/lotus-streaming/blob/master/lib/streaming_hacks.rb.

There are two ways to start the server:

```rackup -s Puma```
```lotus s```

You can see the streaming in action at:

[lotus s](http://localhost:2300)
[rackup -s Puma](http://localhost:9292)

When you open the page in a browser and you see "one", "two", "three" appear one by one, with a pause of 500ms => streaming works. If everything appears after waiting for about 1s => streaming does not work.

As @berfarah has described streaming does not work with ```lotus s``` while it works with ```rackup -s Puma```. 
 What i found is that the problem seems to stem from ```shotgun/ code reloading```, not hijack/rack/chunked or so. If i start the server with 
```lotus s --code_reloading=false``` then streaming works.

@jodosha any ideas? I don't know much about shotgun.

**Comment by pascalbetz** [comment]
Forgot to mention: once i know why ```lotus s``` is not working with code reloading i can start to implement the changes required for streaming to work in a less hackish way.

**Comment by pascalbetz** [comment]
Looked into shotgun. Seems like the problem is here:

https://github.com/rtomayko/shotgun/blob/master/lib/shotgun/loader.rb#L88

The __child__ reads the body (chunks of the stream) and writes it into the ```@writer```, after that it closes the stream and only then the __parent__ returns the body to the client.

AFAICT streaming will not work with shotgun. Since it's planned to remove shotgun anyway (see #249) i'd say i concentrate on making the streaming code provided by @berfarah  a  bit cleaner. If i understand correctly, then Lotus does not want to monkeypatch (@jodosha ?) Suggestions/Feedback welcome.

**Comment by berfarah** [comment]
@pascalbetz I've written a change to the `rendering_policy.rb` for this that will fix this with `rackup`, writing the test now.

**Comment by pascalbetz** [comment]
Nice.

FYI (since i can not comment on the patch):
```ruby
response[BODY] = Array(body) unless body.nil? || body.respond_to?(:each)
```
_nil_ does not respond to each so this could be simplified (might not be easier to read though).

**Comment by berfarah** [comment]
Hey @pascalbetz, I've made some updates on the branch. `#size` isn't a valid method for `Rack::Chunked::Body`, so I've had to resort to uglier methods.

Also, I'm not seeing what you mean about `nil`. Can you elaborate?

https://github.com/lotus/lotus/compare/master...berfarah:feature-streaming

I can add you as a collaborator, if you'd like, on https://github.com/berfarah/lotus/tree/feature-streaming

**Comment by pascalbetz** [comment]
Never mind about the ```nil``` thing. I actually think that, instead of simplifying the boolean expression,  we should move it to it's own method so it has a name.
```def wrap_body?(body)``` or something like that?

**Comment by pascalbetz** [comment]
And: perhaps we can add a note that with shotgun enabled streaming does not work. If you come to the same conclusion about shotgun:-)

**Comment by berfarah** [comment]
I can confirm that disabling shotgun worked for me. However, I'm still struggling to get my [tests](https://github.com/berfarah/lotus/blob/feature-streaming/test/integration/streaming_test.rb) to produce a streaming outcome in [my branch](https://github.com/berfarah/lotus/tree/feature-streaming)

**Comment by jodosha** [comment]
@berfarah @pascalbetz You're awesome! :tophat:

First of all: the check. What @berfarah did is right, we need to introspect the body and decide what to do.
There is a conceptual duplication, because we do the same both in `#render` and `#view_for`. So @pascalbetz suggestion to wrap the body is great.

Now, I'd just introduce a new type `Body`, that does all that checks internally and exposes an API that is meaningful to Rack and `RenderingPolicy`. In other words:

```ruby
def render(env, response)
  # ...
  response[BODY] = Body.new(body)
end
```

Then we can remove the check introduced in `#view_for`, because `Body` responds to `#empty?`.

This is only conceptually right, I don't know if Rack complains about this new type or what.. but I think it's a viable solution.

**Comment by pascalbetz** [comment]
@jodosha 
This should work, AFAIK rack only requires ```each``` to work. By adding ```empt?``` this new class can encapsulate everything. Could look like this (lunchbreak coding):

```ruby

class BodyWrapper
  def initialize(body)
    @body = body
  end

  def empty?
    @body.respond_to?(:empty?) && @body.empty?
  end

  def each(*args, &block)
    wrapped_body.each(*args, &block)
  end

  private

  def wrapped_body
    if @body.respond_to?(:each)
      @body
    else
      Array(@body)
    end
  end

end
```

and then in StreamingPolicy:

```ruby
    def render(env, response)
      response[BODY] = ::BodyWrapper.new(response[BODY])
      body = _render(env, response)
      response[BODY] = ::BodyWrapper.new(body)
      response
    end
```

Which is not yet perfect because we need to wrap the body twice, once before passing it to ```_render``` and once before passing the result back to the caller.

@berfarah 

FYI: the ```use ::Rack::Chunked``` in the action is not needed for me, do you think it's needed?

**Comment by jodosha** [comment]
@pascalbetz I was thinking to:

```ruby
require 'delegate'

class Body < SimpleDelegator
  def initalize(body)
    @body = body.respond_to?(:each) ? body : Array(body)
  end

  def empty?
    @body.respond_to?(:empty?) && @body.empty?
  end
end
```

We just need to check if Rack doesn't get confused by `SimpleDelegator`.

Instead of wrap twice in `#render`, why we don't just make `_render` to return an instance of `Body`, so we can just assign it once?

**Comment by pascalbetz** [comment]
@berfarah i don't like delegators. They make it hard to debug and understand whats going on. Especially since Rack only needs ```each``` as interface.

The other part (not wrapping twice) sounds cool.

**Comment by jodosha** [comment]
@pascalbetz Ok to remove the delegator part, but I think the wrapping must happen on object initialization.

```ruby
class Body
  def initalize(body)
    @body = body.respond_to?(:each) ? body : Array(body)
  end

  def empty?
    @body.respond_to?(:empty?) && @body.empty?
  end

  def each(&blk)
    @body.each(&blk)
  end
end
```

I've noticed that `Rack::Chunked::Body` responds to `#close` as well. It's probably a good idea to implement it too. :+1:

**Comment by berfarah** [comment]
@jodosha @pascalbetz If I wrap the body in `_render` then I run into method missing errors with the default template. Is this abstraction worth it? It seems like we might be over-complicating the code trying to simplify it.

--- 

@pascalbetz https://github.com/lotus/lotus/issues/303#issuecomment-149853341 around `::Rack::Chunked`, I still need that in the application or `lotus s` won't chunk the response. This is because `lotus s` removes `::Rack::Chunked` [here](https://github.com/lotus/lotus/blob/master/lib/lotus/commands/server.rb#L36)

---

I'm still having trouble actually creating a streaming server in tests. My integration test still refuses to actually return a chunked response, despite my being able to run `rackup` and `lotus s` in the fixture and get what I'm looking for. I think the issue lies somewhere with the way rack's test server works, but I haven't found anything conclusive.

Can either of you have a look at the test and given me feedback on what might be going wrong?

https://github.com/lotus/lotus/compare/master...berfarah:feature-streaming

(on a final note, I think I'll remove the `sleep 0.5s` from the test so it doesn't slow down the suite)

**Comment by pascalbetz** [comment]
@jodosha I know you use SimpleDelegator (seen it in the code in other places). I think it would be overkill here and as mentioned it also makes debugging/searching more complicated.

Wrapping on initialization sounds good. We must add close as well, something like this?

```ruby
def close
  @body.close if @body.respond_to?(:close)
end
```
FYI: here is the spec of BODY: http://www.rubydoc.info/github/rack/rack/master/file/SPEC#The_Body



@berfarah then let's get it to a state where it's working and tested and then we can refactor/improve it.
For the test, we should remove the timeout, yes. For playing around i need it in order to see streaming working:-)
I will look at the test case this evening and get back to you if i have anything to add.
If you create PR, then we can discuss there and give feedback on specific lines of code. Not sure how @jodosha and the core team handles this but i like PRs for discussion:-)

About the chunked middleware. Need to test that again in that case.

**Comment by berfarah** [comment]
@pascalbetz  I can do a PR request, I just didn't want to submit a broken test. What do you think, should I submit it anyway?

**Comment by jodosha** [comment]
Solved by #375
